src/libopensc/card-itacns.c

```diff
@@ -54,141 +54,80 @@ static struct sc_card_driver itacns_drv = {
 static const struct sc_atr_table itacns_atrs[] = {
 	{ "3b:f4:18:00:ff:81:31:80:55:00:31:80:00:c7", NULL, NULL,
 		SC_CARD_TYPE_ITACNS_CIE_V1, 0, NULL},
+	{ "3b:8b:80:01:00:31:c1:64:00:00:00:00:00:00:00:00",
+	  "ff:ff:ff:ff:ff:ff:ff:ff:00:00:00:00:00:00:00:00",
+	  "Idemia (Oberthur)", SC_CARD_TYPE_ITACNS_CNS_IDEMIA_2021, 0, NULL},
 	{ NULL, NULL, NULL, 0, 0, NULL}
 };
 
-/* Output debug info */
-#define matchdebug(idx, c) do { \
-	sc_debug(ctx, SC_LOG_DEBUG_VERBOSE, \
-		"Matching %x against atr[%d] == %x", c, idx, atr[idx]); \
-	} while(0);
-
-/* Check that we are not looking at values beyond the ATR's length.
- * If we are, then the card does not match. */
-#define itacns_atr_l(idx) do {if (idx >= card->atr.len) return 0;} while(0);
-
-/* Match byte exactly and increment index. */
-#define itacns_atr_match(idx, c) do { \
-		itacns_atr_l(idx); \
-		matchdebug(idx, c); \
-		if (((u8)atr[idx]) != c) return 0; \
-		idx++; \
-	} while(0);
-
-/* Match masked bits and increment index. */
-#define itacns_atr_mmatch(idx, c, mask) do { \
-		itacns_atr_l(idx); \
-		if ((((u8)atr[idx]) & mask) != c) return 0; \
-		idx ++; \
-	} while(0);
-
 /* Macro to access private driver data. */
 #define DRVDATA(card) ((itacns_drv_data_t *) card->drv_data)
 
-
-static int itacns_match_cns_card(sc_card_t *card, unsigned int i)
+static void itacns_init_cns_card(sc_card_t *card)
 {
-	unsigned char *atr = card->atr.value;
-	sc_context_t *ctx;
-	ctx = card->ctx;
+	if (15 != card->reader->atr_info.hist_bytes_len)
+		return;
 
+	u8 cns_version = card->reader->atr_info.hist_bytes[12];
+	card->version.hw_major = (cns_version >> 4) & 0x0f;
+	card->version.hw_minor = cns_version & 0x0f;
```

Comment:
I'm not sure if this is the correct place to put it. hardware version is about the hardware design of the chip and all its surroundings. software version is about what software (OS) the smart card chip is running. the CNS version, however, is the version of the standard which the chip OS is conforming to, e.g. which APDUs it accepts.